### PR TITLE
Enhancement: better slope selector plots

### LINKED ISSUE #23: Enhancement: Slope Selector Plots Display
**Description**

Currently slope selector plots are ugly. Aim to show one plot per page. If user wants >1 plot per page plots should be in grid format. 

**To do**

- [ ] One slope plot per page
- [ ] Option for Multiple plots on page in grid format
- [ ] Improve axis labels, titles and font sizes etc so plots are easy to read and no overlapping sections or unclear titles,

### PR DESCRIPTION
**Pull request by m-kolomanski**
## Issue

Closes #23, #22 and #36 

## Description

- Split slope selection functionality into separate module
- Improved UI and UX for the slope exclusion functionality
- Fixed help widget for the module
- Introduces `shinyjs`, `reactable` and `reactable.extras` packages to the app
- Refactored code, moved logic into separate functions, streamlined data flow

## Definition of Done

- [x] slope selector functionality migrated to separate module
- [x] addressed complexity of server function for the module
- [x] fixed warnings generated by `ggplot2` and `plotly`
- [x] fixed overlapping annotations on the plots
- [x] fixed/refactored help widget
- [x] improved page layout for plots
- [x] updated gifs in manual widget

## How to test

Open the application and run NCA as normal. Go to Setup -> Slope Selector tab and play around with the plots. You can provide selections and exclusions either by data table or by clicking on the plots themselves. After each change, the plots should update to reflect it. To run the analysis again and apply ruleset to the actual data, the **Apply** button will need to be clicked.

## Contributor checklist
- [x] Code passes lintr checks
- [x] Code passes all unit tests
- [x] New logic covered by unit tests
- [x] New logic is documented

### CODE REVIEW

**Review by js3110 — COMMENTED** [review]
Looks so much better than before! Still some things to do:
message when i run devtools::load_all():
✖ `.check_slope_rule_overlap` masks `aNCA::.check_slope_rule_overlap()`.
ℹ Did you accidentally source a file rather than using `load_all()`?
  Run `rm(list = c(".check_slope_rule_overlap"))` to remove the conflicts.

- Table still doesn't change if you remove an exclusion via plot click
- After I click the same plots a couple of times it is showing all the plots on one page just in different formats depending on how many per page is selected
- Occasionally if I click stuff or add rows and apply it bugs out for a moment or the screen freezes.

**Comment by m-kolomanski** [comment]
Hej @js3110,

sadly, I do not get any of those issues.

For messages about masking, did you try terminating your R session and reloading it completely? Might be some lingering imports that cause those issues.

For me, the table changes in accordance to the plots. The row is also removed when there is no exclusions left.
[Screencast from 2024-11-07 08-38-09.webm](https://github.com/user-attachments/assets/145a13d1-4e2b-46da-80a9-5a7ac7f73662)

Would you mind providing more detailed steps to reproduce those bugs?

**Comment by js3110** [comment]
Hi @m-kolomanski 

I restarted my R and now I don't get the warning message!
The table also works now and removes the rows, however I still have some issues with the screen freezing and the plots showing up. The first time I ran the nca and applied a manual slope the results page froze and didnt show me the updates. I then tried again screen recording and it worked a bit better however it still caused some issues. I have sent you the screen recording on slack (it was too big to copy here)

**Comment by js3110** [comment]
@m-kolomanski @Gotfrid I just tried to do a full check, and the slope selector feature is working amazing now, however there are a couple more bugs now that are probably knock on effects from all the changes made to the slope selector:

- [ ] Partial auc button does not work (Error in partial_auc_input: could not find function "partial_auc_input")
- [ ] Slope Settings upload does not work (Error in if: argument is of length zero)

File: R/utils-slope_selector.R [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    if (length(existing_index) > 1)
+      log_warn("More than one range for single patient, profile and rule type detected.")
+    return(rbind(existing, new))
+  }
+
+  existing_range <- .eval_range(existing$IXrange[existing_index])
+  new_range <- .eval_range(new$IXrange)
+
+  is_inter <- length(intersect(existing_range, new_range)) != 0
+  is_diff <- length(setdiff(new_range, existing_range)) != 0
```
[the remark this answers is not in the document]
Comment (m-kolomanski) [reply R/utils-slope_selector.R:101]: Yes, this is intentional, I want to check if `new_range` introduces any new points to the equation that `existing_range` does not have.

File: R/utils-slope_selector.R [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  data$conc$data$exclude_half.life <- FALSE
+
+  # If there is no specification there is nothing to save #
+  if (nrow(slopes) == 0) {
+    return(data)
+  }
+
+  # Eliminate all rows with conflicting or blank values
+  slopes <- slopes %>%
+    dplyr::filter(
```
[the remark this answers is not in the document]
Comment (m-kolomanski) [reply R/utils-slope_selector.R:30]: Question: do we really need `@importFrom dplyr filter` if we namespace the function call? There is `dplyr::group_by` and `dplyr::mutate` in the same file, should I include those functions too in the import statement?

**Comment by js3110** [comment]
@m-kolomanski the first plot has no height for me

File: R/utils-slope_selector.R [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      TYPE %in% c("Selection", "Exclusion"),
+      PATIENT %in% names(profiles),
+      PROFILE %in% unname(unlist(profiles[PATIENT])),
+      all(!is.na(sapply(IXrange, function(x) .eval_range(x))))
+    ) %>%
+    # Eliminate duplicated records within the same profile
+    dplyr::filter(
+      !duplicated(
+        paste0(PATIENT, PROFILE, IXrange, fromLast = TRUE),
```
[the remark this answers is not in the document]
Comment (Gero1999) [reply R/utils-slope_selector.R]: Hello @Gotfrid, you are right the code seems wrong. The original intention of the code is that whenever "Save" is pressed by the user, conflictive records only prioritize the last input in the table.  Let me put a practical example to explain it better:

```
slope_table <- 
  data.frame(TYPE = c("Selection", "Selection", "Exclusion", "Exclusion"), 
             PATIENT = c("S1", "S1", "S1", "S1"),
             PROFILE = c(1,1,1,1), 
             IXrange = c("1:3", "2:4", "3", "3")
             )

slope_table_filt <- slope_table %>% 
  # Eliminate duplicated records (i.e, duplicated inclusions or, in this case, exclusions)
  dplyr::filter(!duplicated(paste0(TYPE, PATIENT, PROFILE, IXrange))) %>%
  
  # Eliminate conflictive selections, only 1 per profile (i.e, 1:3 & 2:4)
  dplyr::filter(TYPE == "Exclusion" | (TYPE == "Selection" & duplicated(PATIENT, PROFILE, fromLast = TRUE)))
```

if you find a better solution to perform the same action with `distinct` or any other way, please feel free to modify it. I hope that now with the correction made at least is clear the purpose of the action, if not let me know!

**Review by js3110 — APPROVED** [review]
Nice!!

File: R/utils-slope_selector.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (js3110) [reply R/utils-slope_selector.R]: @m-kolomanski I think if they change it then its best to keep the latest one as its likely they changed it for a reason. If they wanted 1:4 they would select that in the first place.

**Review by Gotfrid — APPROVED** [review]
LGTM

**Pull request merged**